### PR TITLE
[SYCL-MLIR][LICM] Continue to check the other operations for conflict

### PR DESCRIPTION
**Pull request by whitneywhtsang**
We used to call `conflictsWithOperationInBlock` to get the first operation conflicting with the given operation `op`, then allow hoisting if the conflicting operation is in `willBeMoved`, but there could be other operations conflict with `op` after and not in `willBeMoved`.

### CODE REVIEW

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp [hunk 1]
```diff
@@ -96,24 +96,6 @@ class OperationSideEffects {
...
-  /// Returns the first operation in the given region \p rgn with side effects
-  /// that conflict with the side effects summarized in this class, and a
-  /// std::nullopt if none is found.
-  Optional<Operation *> conflictsWithOperationInRegion(Region &rgn) const;
-
-  /// Returns the first operation in the given loop \p loop with side effects
-  /// that conflict with the side effects summarized in this class, and a
-  /// std::nullopt if none is found.
-  Optional<Operation *>
-  conflictsWithOperationInLoop(LoopLikeOpInterface loop) const;
```
Comment (whitneywhtsang) [inline polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:115]: `conflictsWithOperationInRegion` and `conflictsWithOperationInLoop` are not used.

File: polygeist/test/polygeist-opt/sycl/matrix_multiply_reduction.mlir [hunk 2]
```diff
@@ -26,7 +26,7 @@
 !sycl_accessor_write_1_ = !sycl.accessor<[1, f32, write, global_buffer], (!sycl_accessor_impl_device_1_, !llvm.struct<(memref<?xf32, 1>)>)>
 !sycl_accessor_read_1_ = !sycl.accessor<[1, f32, read, global_buffer], (!sycl_accessor_impl_device_1_, !llvm.struct<(memref<?xf32, 1>)>)>
 
-func.func private @matrix_multiply_reduction(%arg0: memref<?x!llvm.struct<(!sycl_accessor_write_1_, !sycl_accessor_read_1_, !sycl_accessor_read_1_)>, 4>, %arg1: i32) {
+func.func private @matrix_multiply_reduction(%arg0: memref<?x!sycl_accessor_write_1_, 4>, %arg1: memref<?x!sycl_accessor_read_1_, 4>, %arg2:memref<?x!sycl_accessor_read_1_, 4>, %arg3: i32) {
```
Comment (whitneywhtsang) [inline polygeist/test/polygeist-opt/sycl/matrix_multiply_reduction.mlir:29]: After fixing the problem, some operations are unable to be hoisted. By splitting the memref struct argument like `ArgumentPromotion`, they can then be hoisted.

File: polygeist/tools/cgeist/driver.cc [hunk 3]
```diff
@@ -361,13 +361,17 @@ static int optimize(mlir::MLIRContext &Ctx,
...
   enableOptionsPM(PM);
 
-  mlir::OpPassManager &OptPM = PM.nestAny();
   GreedyRewriteConfig CanonicalizerConfig;
   CanonicalizerConfig.maxIterations = CanonicalizeIterations;
 
   if (OptLevel != llvm::OptimizationLevel::O0) {
     PM.addPass(polygeist::createArgumentPromotionPass());
 
+    mlir::OpPassManager &OptPM = PM.nestAny();
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/driver.cc:370]: LICM needs to be after ArgumentPromotion, in order to catch the matrix mulitply reduction opportunity.

File: polygeist/test/polygeist-opt/sycl/matrix_multiply_reduction.mlir [hunk 4]
```diff
@@ -1,4 +1,4 @@
-// RUN: polygeist-opt -licm -raise-scf-to-affine -detect-reduction %s | FileCheck %s
+// RUN: polygeist-opt -arg-promotion -licm -raise-scf-to-affine -detect-reduction %s | FileCheck %s
```
Comment (whitneywhtsang) [inline polygeist/test/polygeist-opt/sycl/matrix_multiply_reduction.mlir:1]: After fixing the problem, some operations are unable to be hoisted. By splitting the memref struct argument, they can then be hoisted.

**Comment by whitneywhtsang** [comment]
Checked that the matrix multiple reduction microbenchmark is good. However, there are regressions on SYCL-Bench, need to investigate why we are no longer catching some of the opportunities in SYCL-Bench.

**Review by victor-eds — APPROVED** [review]

File: polygeist/tools/cgeist/driver.cc [hunk 5]
```diff
@@ -361,13 +361,17 @@ static int optimize(mlir::MLIRContext &Ctx,
...
+    if (EnableLICM)
+      OptPM.addPass(polygeist::createLICMPass(
+          {options.getCgeistOpts().getRelaxedAliasing()}));
```
Comment (victor-eds) [inline polygeist/tools/cgeist/driver.cc:373]: Maybe the pipeline change should be performed in a different PR to see if this is where this pass should go.

**Review by victor-eds — CHANGES_REQUESTED** [review]
Change LGTM but for pipeline change

File: polygeist/test/polygeist-opt/sycl/matrix_multiply_reduction.mlir (hunk 4 above)
Comment (etiotto) [reply polygeist/test/polygeist-opt/sycl/matrix_multiply_reduction.mlir:1]: I think we need to understand what operation prevent LICM to hoist code out of the loop (in the absence of arg-promotion). I am just worry that arg-promotion hides another problem.

File: polygeist/test/polygeist-opt/sycl/matrix_multiply_reduction.mlir [hunk 6]
```diff
@@ -1,4 +1,4 @@
...
 
 // CHECK-LABEL: func.func private @matrix_multiply_reduction
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply polygeist/test/polygeist-opt/sycl/matrix_multiply_reduction.mlir:3]: `polygeist.subindex` has the `ViewLikeOpInterface`, which means for `%0 = polygeist.subindex(%arg0, %c0)`, when determining if `%0` alias with another operation, `%arg0` is used.
`%arg0` is a memref of struct of accessors, which doesn't satisfy the check in `sycl::AliasAnalysis::handleAccessorSubscriptAlias`.

After argument promotion, uses of `%0` is changed to the new `%arg0`. The new `%arg0` has type memref of accessor, which satisfy the check in `sycl::AliasAnalysis::handleAccessorSubscriptAlias`.

**Comment by whitneywhtsang** [comment]
> Checked that the matrix multiple reduction microbenchmark is good. However, there are regressions on SYCL-Bench, need to investigate why we are no longer catching some of the opportunities in SYCL-Bench.

Regression on SYCL-Bench (with workaround) is fixed with https://github.com/intel/llvm/pull/8893.

File: polygeist/test/polygeist-opt/sycl/matrix_multiply_reduction.mlir (hunk 4 above)
Comment (whitneywhtsang) [reply polygeist/test/polygeist-opt/sycl/matrix_multiply_reduction.mlir:1]: Answered in https://github.com/intel/llvm/pull/8872/files#r1153668918.

**Comment by whitneywhtsang** [comment]
The L0 GEN9 LLVM Test Suite failure is not real, as `image_max_size.cpp` is flaky, it sometimes time out, and sometimes not.
Created https://github.com/intel/llvm/pull/8899 to mark that test case as unsupported on L0.

**Comment by etiotto** [comment]
https://github.com/intel/llvm/pull/8899 has landed.

**Review by etiotto — APPROVED** [review]
LGTM. We should merge this PR in once the PRs needed to get the performance opportunities in SYCL-Bench have landed.

**Review by victor-eds — APPROVED** [review]
